tools/png-to-badge-asset.1
.TH PNG-TO-BADGE-ASSET "1" "APRIL 2024" "png-to-badge-asset" "User Commands"
.SH NAME
png-to-badge-asset \- generate C code from PNG files for use with the badge
.SH SYNOPSIS
.B png-to-badge-asset [-s] N png-file ...

.SH DESCRIPTION
.\" Add any additional description here
.PP
png-to-badge-asset generates C code representations of PNG files suitable
for use with the badge firmware.  The value of N should be one of 16, 8,
4, 2, or 1, representing the type of image data you wish to generate,
described below.
.PP
16-bit images: each pixel is represented by 16 bits encoding the color directly
with 5 bits each for red, green, and blue. Transparency is not supported.
.PP
8,4 and 2 bit images: a color map is generated supporting up to 255, 15, or 3
colors, plus transparency, respectively, and for each pixel the color is
represented by an 8, 4 or or 2 bit index into the aforementioned color table.
.PP
1 bit images: No color table is generated, and each pixel is represented by
a single bit.  When drawn, the image colors will be determined by the current
foreground and background colors, as set by FbColor() and FbBackgroundColor().
.PP
The C code is sent to stdout so you can redirect it into a file, or capture
it with the mouse for cutting and pasting into your code.  Diagnostics and
status messages are sent to stderr.
.SH OPTIONS
.TP
\fB\-s, \-\-nostatic\fR
By default the C code representing the images which
png-to-badge-asset produces declares all variables with
static scope.  That's fine in the usual case where only a
single .c file needs access to them.  In some cases though
multiple .c files need access and static scope is not
appropriate.  The --nostatic option suppresses the static
keyword.

Additionally, the definitions are surrounded by the conditional
compilation symbol DEFINE_IMAGE_ASSET_DATA and if this is not
defined, only an extern declaration is compiled. That is to say,
the same code may be included in one place to define the data
and in all others to declare the extern declaration.
.SH WARNINGS AND ERRORS
.PP
Errors may occur in the following instances:
.PP
Too many colors: The PNG file contains too many colors for the chosen image format.
Choosing an output image format that supports more colors, or posterizing the image
to reduce the number of colors present may help.
.PP
Image too large: If you process images larger than the display of the badge (128 x 160)
a warning will be issued (but the image will still be processed.)
.PP
Transparency ignored: If you process a PNG file which contains an alpha channel
into the 1 or 16 bit images formats, a warning will be issued as these formats
do not support transparency. The image will still be processed.
.SH EXAMPLE
.TP
.DS
.sp
.nf
$ ../../tools/png-to-badge-asset --nostatic 8 badgey_assets/star.png 
Processing badgey_assets/star.png
/* Begin code generated by png-to-badge-asset from badgey_assets/star.png Thu Apr 18 10:44:15 2024. */
#ifdef DEFINE_IMAGE_ASSET_DATA
static const uint16_t star_colormap[51] = {
	0x1081, 0x0000, 0xa4cd, 0x5286, 0x3143, 0x6b08, 0x2902, 0x0800, 0x39c4, 0x3184, 0x3984, 0x18c1,
	0x4204, 0x39c5, 0x5246, 0x62c7, 0x5ac7, 0x5a86, 0x7348, 0xa4cc, 0xa48c, 0x83c9, 0x2943, 0xc5ce,
	0xe690, 0xeed0, 0x83ca, 0x9c4b, 0xc58e, 0xde90, 0xe6d0, 0xf757, 0x944b, 0xeed1, 0xf711, 0xff57,
	0xad0e, 0x944c, 0x2102, 0x4a46, 0xde4f, 0x1881, 0x940a, 0x7b89, 0x5a87, 0x41c4, 0x6b07, 0x2903,
	0x4a05, 0x2942, 0x0840,
}; /* 51 values */

static const uint8_t star_data[256] = {
	0, 0, 1, 1, 1, 1, 1, 2, 3, 1, 1, 1, 1, 1, 1, 1, 1, 4, 1, 1,
	1, 1, 1, 2, 5, 1, 1, 1, 1, 1, 6, 7, 1, 0, 8, 1, 1, 1, 1, 9,
	2, 1, 1, 1, 1, 10, 11, 1, 1, 1, 12, 13, 1, 1, 1, 3, 2, 1, 1, 1,
	14, 8, 1, 1, 1, 1, 1, 15, 16, 1, 1, 2, 2, 1, 2, 2, 17, 1, 1, 1,
	1, 1, 1, 2, 2, 18, 2, 2, 19, 2, 20, 21, 1, 1, 1, 1, 1, 1, 1, 1,
	22, 23, 2, 24, 25, 23, 23, 1, 1, 1, 1, 1, 26, 2, 2, 27, 28, 29, 30, 31,
	31, 23, 23, 2, 2, 32, 2, 2, 2, 23, 2, 23, 23, 33, 34, 35, 31, 24, 2, 36,
	37, 2, 2, 5, 1, 38, 39, 2, 2, 2, 29, 25, 25, 40, 2, 17, 41, 10, 11, 1,
	1, 1, 1, 3, 3, 23, 2, 23, 40, 23, 23, 2, 1, 1, 1, 1, 1, 1, 1, 9,
	2, 16, 2, 2, 23, 42, 19, 5, 1, 1, 1, 1, 1, 1, 8, 43, 2, 44, 45, 2,
	2, 10, 8, 46, 1, 1, 1, 1, 1, 47, 48, 1, 1, 1, 1, 2, 39, 1, 1, 45,
	8, 1, 1, 1, 49, 4, 1, 1, 1, 1, 1, 2, 16, 1, 1, 1, 1, 1, 1, 1,
	50, 1, 1, 1, 1, 1, 1, 15, 2, 1, 1, 1, 1, 1, 1, 1,
}; /* 256 values */

const struct asset2 star = {
	.type = PICTURE8BIT,
	.seqNum = 1,
	.x = 16,
	.y = 16,
	.colormap = (const uint16_t *) star_colormap,
	.pixel = (const unsigned char *) star_data,
};
#else
extern const struct asset2 star;
#endif
/* End of code generated by png-to-badge-asset from badgey_assets/star.png Thu Apr 18 10:44:15 2024 */
Processed 2 file(s) with 0 error(s) and 0 warning(s).
.fi
.DE
.sp
.PP
.SH AUTHOR
png-to-badge-asset was written by Stephen M. Cameron
.br


CHANGELOG.txt
(Not a true CHANGELOG, this is a running list of notes about changes in the new_badge_monsters branch)

1. Added 128x128 version of all Bill's creations in /new_badge_monster_assets/ directory.
2. Refactored `tools/asset_converter.py`, fixed some apparent bugs, and generated files for all assets.
3. added script, configuration file for `cpplint` linting of C code, cleaned up some issues.
4. refactored `badge_monsters.c/h` into the /new_badge_monsters/ sub-directory, split out IR functions
into a separate file.
5. Revised data structures and code to support use of `FbImage4Bit` for drawing monsters.


tools/png-to-badge-asset.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <time.h>
#include <getopt.h>

#include "../source/hal/png_utils.h"

#define LCD_XSIZE 128
#define LCD_YSIZE 160

#define TRANSPARENT_INDEX 255
#define ALPHA_THRESHOLD 128

static int errors = 0;
static int warnings = 0;
static uint16_t colormap[256] = { 0 };
static int ncolors = 0;
#define MAX_OVERFLOW_COLORS (0x0ffff)
static uint16_t overflow_colormap[MAX_OVERFLOW_COLORS];
static int noverflow_colors;
static int maxcolors = 256;
static uint8_t *pixdata = NULL;
static uint16_t *pixdata16 = NULL;
static int pixdata_size = 0;
static int pixels_per_byte = 1;
static int pixel_count = 0;
static int bits_per_pixel = 8;
static int bits_per_pixel_mask = 0xff;
static int color_count = 0;
static int nostatic = 0;
static int skipargs = 0;
char *static_keyword = "static ";

static void __attribute__((noreturn)) usage(char *program_name)
{
	char *p = program_name ? program_name : "png-to-badge-asset";

	fprintf(stderr, "%s: usage: %s [--nostatic] bits-per-pixel png-file(s)\n", p, p);
	fprintf(stderr, "      bits-per-pixel may be 16, 8, 4, 2, or 1\n");
	exit(1);
}

/* Returns the index into the colormap of the pixel color, possibly adding
 * the color to the colormap.  If there is no room left in the colormap,
 * -1 is returned.  Colors are stored in the colormap in 16 bit quantites,
 * with 5 bits each for R, G, and B with R in bits 11 - 15, G in bits 6-10,
 * and B in bits 0-4.  Bit 5 is unused.
 */
static int add_to_colormap(unsigned char *pixel)
{
	/* Compute the 16-bit version of the color */
	uint16_t color;
	uint8_t r, g, b;

	r = pixel[0];
	g = pixel[1];
	b = pixel[2];

	color = ((((r >> 3) & 0b11111) << 11)
		| (((g >> 3) & 0b11111) <<  6)
		| (((b >> 3) & 0b11111)));

	/* Check if we've already stored this color in cmap */
	for (int i = 0; i < ncolors; i++) {
		if (colormap[i] == color)
			return i;
	}

	/* Not found, add the new color */
	if (ncolors < maxcolors) {
		colormap[ncolors] = color;
		ncolors++;
		color_count++;
		return (ncolors - 1);
	}

	/* No room left in colormap, add to overflow_colormap.
	 * This just so we can accurately count and report the actual
	 * number of colors in the image data.  We don't actually do
	 * anything with the overflow colors.
	 */
	for (int i = 0; i < noverflow_colors; i++) {
		if (overflow_colormap[i] == color)
			return -1;
	}

	if (noverflow_colors < MAX_OVERFLOW_COLORS) {
		overflow_colormap[noverflow_colors] = color;
		noverflow_colors++;
		color_count++; 
	}
	
	return -1;
}

/* Calculate bytes per row for decoded png data. If alpha channel is present, it's easy
 * just 4 bytes * pixels per row.  If no alpha channel, it's 3 bytes * pixels per row,
 * plus possible a few more bytes to make things evenly divisible by 4.
 */
static int calculate_bytes_per_row(int width, int hasalpha)
{
	int answer;

	if (hasalpha)
		return width * 4;
	answer = width * 3;
	while (answer & 0x03)
		answer++;
	return answer;
}

static void free_pixdata(void)
{
	if (pixdata)
		free(pixdata);
	if (pixdata16)
		free(pixdata16);
	pixdata = NULL;
	pixdata16 = NULL;
}

/* Generate the color map (list of RGB colors used in the image, packed into 16 bit quantities)
 * and the pixel data (one value per pixel, indexing into the color map.)  The pixel data can
 * be packed into bytes in quanties of 1, 2, 4, or 8 pixels per byte, depending on pixels_per_byte
 */
static int generate_colormap_and_pixdata(char *image, int width, int height, int hasalpha)
{
	unsigned char *img = (unsigned char *) image;
	unsigned char *pixel;
	unsigned char alpha;
	int bytes_per_pixel = hasalpha ? 4 : 3;
	int bytes_per_row = calculate_bytes_per_row(width, hasalpha);

	color_count = 0;
	pixdata_size = (width * height) / pixels_per_byte;
	pixdata = malloc(pixdata_size);
	if (!pixdata) {
		fprintf(stderr, "  Error: Failed to allocate %d bytes\n", pixdata_size);
		return -1;
	}
	memset(pixdata, 0, pixdata_size);
	pixel_count = 0;

	for (int y = 0; y < height; y++) {
		unsigned char *start_of_row = img + (y * bytes_per_row);
		for (int x = 0; x < width; x++) {
			int index;
			pixel = start_of_row + x * bytes_per_pixel;
			if (hasalpha && bits_per_pixel != 16 && bits_per_pixel != 1)
				alpha = pixel[3];
			else
				alpha = 255;
			if (alpha < 128) {
				index = TRANSPARENT_INDEX & bits_per_pixel_mask;
			} else {
				index = add_to_colormap(pixel);
			}
			if (index < 0)
				continue;
			int byte = pixel_count / pixels_per_byte;
			int shift = pixels_per_byte - (pixel_count % pixels_per_byte) - 1;
			pixdata[byte] |= (index << (shift * bits_per_pixel));
			pixel_count++;
		}
	}
	if (color_count > maxcolors) {
		fprintf(stderr, "  Error: Too many colors (%d) in image (max = %d)\n",
			color_count, maxcolors);
		fprintf(stderr, "  You should posterize the image to reduce colors.\n");
		return -1;
	}
	return 0;
}

/* Generate the 16-bit pixel data (one 16-bit value per pixel color data stored in pixdata16)
 */
static int generate_pixdata16(char *image, int width, int height, int hasalpha)
{
	unsigned char *img = (unsigned char *) image;
	unsigned char *pixel;
	int bytes_per_pixel = hasalpha ? 4 : 3;
	int bytes_per_row = calculate_bytes_per_row(width, hasalpha);
	pixdata_size = 2 * width * height;
	pixdata16 = malloc(pixdata_size);
	if (!pixdata16) {
		fprintf(stderr, "  Error: Failed to allocate %d bytes\n", pixdata_size);
		return -1;
	}
	memset(pixdata16, 0, pixdata_size);
	pixel_count = 0;

	for (int y = 0; y < height; y++) {
		unsigned char *start_of_row = img + (y * bytes_per_row);
		for (int x = 0; x < width; x++) {
			pixel = start_of_row + x * bytes_per_pixel;
			/* Compute the 16-bit version of the color */
			uint16_t color;
			uint8_t r, g, b;

			r = pixel[0];
			g = pixel[1];
			b = pixel[2];

			color = ((((r >> 3) & 0b11111) << 11)
				| (((g >> 3) & 0b11111) <<  6)
				| (((b >> 3) & 0b11111)));

			pixdata16[pixel_count] = color;
			pixel_count++;
		}
	}
	return 0;
}

/* Generate the 1-bit pixel data (8 pixels per byte, stored in pixdata, no color map indexing)
 */
static int generate_pixdata1(char *image, int width, int height, int hasalpha)
{
	unsigned char *img = (unsigned char *) image;
	unsigned char *pixel;
	int bytes_per_pixel = hasalpha ? 4 : 3;
	int bytes_per_row = calculate_bytes_per_row(width, hasalpha);
	int output_bytes_per_row = width / 8;
	if (8 * output_bytes_per_row < width) {
		output_bytes_per_row++;
	}
	pixdata_size = (output_bytes_per_row * height);
	pixdata = malloc(pixdata_size);
	if (!pixdata) {
		fprintf(stderr, "  Error: Failed to allocate %d bytes\n", pixdata_size);
		return -1;
	}
	memset(pixdata, 0, pixdata_size);
	pixel_count = 0;
	int byte = 0;
	int bit = 0;
	for (int y = 0; y < height; y++) {
		unsigned char *start_of_row = img + (y * bytes_per_row);
		byte = y * output_bytes_per_row;
		bit = 0;
		for (int x = 0; x < width; x++) {
			pixel = start_of_row + x * bytes_per_pixel;
			/* Compute the 1-bit version of the color */
			uint16_t color;
			uint8_t r, g, b;

			r = pixel[0];
			g = pixel[1];
			b = pixel[2];

			color = r + g + b;
			if (color > 127 * 3)
				color = 1;
			else
				color = 0;

			pixdata[byte] |= (color << bit);
			pixel_count++;
			bit++;
			if (bit > 7) {
				bit = 0;
				byte++;
			}
		}
	}
	return 0;
}


static char *after_last_slash(char *name)
{
	char *last_slash = name;
	for (int i = 0; name[i] != '\0'; i++) {
		if (name[i] == '\\' || name[i] == '/') {
			if (name[i + 1] != '\\' && name [i + 1] != '/')
				last_slash = &name[i + 1];
		}
	}
	return last_slash;
}

/* Generate C code for the color map, pixel data, and asset structures needed
 * by the badge drawing routines.
 */
static int generate_c_code(char *filename, char *prefix, int width, int height)
{
	char *tmp_image_name = strdup(filename);
	char *dot = strstr(tmp_image_name, ".png");
	if (!dot)
		dot = strstr(tmp_image_name, ".PNG");
	if (dot)
		*dot = '\0'; /* strip off the ".png" or ".PNG" from the name */

	/* if there are leading directory names, get rid of them */
	char *image_name = after_last_slash(tmp_image_name);

	/* convert dashes to underscores */
	for (char *i = image_name; *i; i++)
		if (*i == '-')
			*i = '_';


	time_t lt = time(NULL);
	struct tm* now = localtime(&lt);
	char *when = strdup(asctime(now));

	/* asctime puts a newline on the end of the time for some reason.  Cut it off. */
	char *newline = strstr(when, "\n");
	if (newline)
		*newline = '\0';

	printf("/* Begin code generated by png-to-badge-asset from %s %s. */\n", filename, when);
	if (bits_per_pixel != 16 && bits_per_pixel != 1) {
		if (nostatic)
			printf("#ifdef DEFINE_IMAGE_ASSET_DATA\n");
		printf("static const uint16_t %s%s_colormap[%d] = {\n", prefix, image_name, ncolors);
		printf("\t");
		for (int i = 0; i < ncolors; i++) {
			if ((i % 12) == 11)
				printf("0x%04hx,\n\t", colormap[i]);
			else
				printf("0x%04hx,%s", colormap[i], i == ncolors - 1 ? "" : " ");
		}
		printf("\n}; /* %d values */\n", ncolors);
		printf("\n");
		int datacount = pixel_count;
		switch (bits_per_pixel) {
		case 8: datacount = pixel_count;
			break;
		case 4: datacount = pixel_count / 2;
			if ((datacount * 2) < pixel_count)
				datacount++;
			break;
		case 2: datacount = pixel_count / 4;
			if ((datacount * 4) < pixel_count)
				datacount++;
			break;
		default: /* should not be reachable */
			fprintf(stderr, "BUG at %s:%d -- hit unreachable code\n",
				__FILE__, __LINE__);
			exit(1);
			break;
		}
		printf("static const uint8_t %s%s_data[%d] = {\n", prefix, image_name, datacount);
		printf("\t");
		for (int i = 0; i < datacount; i++) {
			if ((i % 20) == 19)
				printf("%hhu,\n\t", pixdata[i]);
			else
				printf("%hhu,%s", pixdata[i], i == pixel_count - 1 ? "" : " ");
		}
		printf("\n}; /* %d values */\n", datacount);
	} else if (bits_per_pixel == 16) {
		printf("static const uint16_t %s%s_data[%d] = {\n", prefix, image_name, pixel_count);
		printf("\t");
		for (int i = 0; i < pixel_count; i++) {
			if ((i % 12) == 11)
				printf("0x%04hx,\n\t", pixdata16[i]);
			else
				printf("0x%04hx,%s", pixdata16[i], i == pixel_count - 1 ? "" : " ");
		}
		printf("\n}; /* %d values */\n", pixel_count);
	} else if (bits_per_pixel == 1) {
		int bytes_per_row = width / 8;
		if (bytes_per_row * 8 < width)
			bytes_per_row++;
		int datacount = bytes_per_row * height;
		if ((datacount * 8) < pixel_count)
			datacount++;
		printf("static const uint8_t %s%s_data[%d] = {\n", prefix, image_name, datacount);
		printf("\t");
		for (int i = 0; i < datacount; i++) {
			if ((i % 16) == 15)
				printf("0x%02hhx,\n\t", pixdata[i]);
			else
				printf("0x%02hhx,%s", pixdata[i], i == datacount - 1 ? "" : " ");
		}
		printf("\n}; /* %d values */\n", datacount);
	}
	printf("\n");
	printf("%sconst struct asset2 %s%s = {\n", static_keyword, prefix, image_name);
	printf("\t.type = PICTURE%dBIT,\n", bits_per_pixel);
	printf("\t.seqNum = 1,\n");
	printf("\t.x = %d,\n", width);
	printf("\t.y = %d,\n", height);
	if (bits_per_pixel != 16 && bits_per_pixel != 1) {
		printf("\t.colormap = (const uint16_t *) %s%s_colormap,\n", prefix, image_name);
		printf("\t.pixel = (const unsigned char *) %s%s_data,\n", prefix, image_name);
	} else if (bits_per_pixel == 16) {
		printf("\t.colormap = NULL, /* not used by PICTURE16BIT or PICTURE1BIT */\n");
		printf("\t.pixel16 = (const uint16_t *) %s%s_data,\n", prefix, image_name);
	} else if (bits_per_pixel == 1) {
		printf("\t.colormap = NULL, /* not used by PICTURE16BIT or PICTURE1BIT */\n");
		printf("\t.pixel = (const uint8_t *) %s%s_data,\n", prefix, image_name);
	}
	printf("};\n");

	if (nostatic) {
		printf("#else\n");
		printf("extern const struct asset2 %s%s;\n", prefix, image_name);
		printf("#endif\n");
	}
	printf("/* End of code generated by png-to-badge-asset from %s %s */\n", filename, when);

	free(when);
	free(tmp_image_name);

	return 0;
}

/* Process a png image turning it into a badge asset implemented in C code */
static int process_image(char *filename, char *prefix)
{
	char *image;
	int width, height, hasalpha;
	char whynot[1000];

	fprintf(stderr, "Processing %s\n", filename);

	width = 0;
	height = 0;
	hasalpha = 0;
	ncolors = 0;
	image = png_utils_read_png_image(filename, 0, 0, 0,
				&width, &height, &hasalpha, whynot, sizeof(whynot));

	if (!image) {
		fprintf(stderr, "  Error processing %s: %s\n",
			filename, whynot);
		return 1;
	}
	if (width > LCD_XSIZE) {
		fprintf(stderr, "  Warning: %s x dimension %d > %d\n", filename, width, LCD_XSIZE);
		warnings++;
	}
	if (height > LCD_YSIZE) {
		fprintf(stderr, "  Warning: %s y dimension %d > %d\n", filename, height, LCD_YSIZE);
		warnings++;
	}
	if (hasalpha && (bits_per_pixel == 16 || bits_per_pixel == 1)) {
		fprintf(stderr, "  Warning: %s has an alpha channel which will be ignored\n", filename);
		warnings++;
	}

	if (bits_per_pixel == 16) {
		int rc = generate_pixdata16(image, width, height, hasalpha);
		if (rc < 0)
			goto error_out;
	} else if (bits_per_pixel == 1) {
		int rc = generate_pixdata1(image, width, height, hasalpha);
		if (rc < 0)
			goto error_out;
	} else {
		int rc = generate_colormap_and_pixdata(image, width, height, hasalpha);
		if (rc < 0)
			goto error_out;
	}
	generate_c_code(filename, prefix, width, height);

	free_pixdata();	
	free(image);
	return 0;

error_out:
	free_pixdata();	
	free(image);
	return 1;
}

static struct option long_options[] = {
	{ "nostatic", no_argument, NULL, 's'}, 
	{ 0, 0, 0, 0 },
};

static void process_options(int argc, char *argv[])
{
	while (1) {
		int option_index;
		int c = getopt_long(argc, argv, "s", long_options, &option_index);
		if (c == -1)
			break;
		switch (c) {
		case 's':
			nostatic = 1;
			static_keyword = "";
			skipargs++;
			break;
		default:
			usage(argv[0]);
			__builtin_unreachable();
			break;
		}
	}
}

int main(int argc, char *argv[])
{

	if (argc < 3) {
		usage(argv[0]);
		__builtin_unreachable();
	}

	process_options(argc, argv);

	int rc = sscanf(argv[1 + skipargs], "%d", &bits_per_pixel);
	if (rc != 1) {
		usage(argv[0]);
		__builtin_unreachable();
	}

	switch (bits_per_pixel) {
	case 1: maxcolors = 2; /* transparency can't be supported */
		pixels_per_byte = 8;
		bits_per_pixel_mask = 0x01;
		break;
	case 2: maxcolors = 3; /* leave one for transparency */
		pixels_per_byte = 4;
		bits_per_pixel_mask = 0x03;
		break;
	case 4: maxcolors = 15; /* leave one for transparency */
		pixels_per_byte = 2;
		bits_per_pixel_mask = 0x0f;
		break;
	case 8: maxcolors = 255; /* save one for transparent */
		pixels_per_byte = 1;
		bits_per_pixel_mask = 0xff;
		break;
	case 16:maxcolors = 0x0ffff;
		pixels_per_byte = 1; /* unused */
		bits_per_pixel_mask = 0x0ffff;
		break;
	default:
		usage(argv[0]);
		__builtin_unreachable();
		break;
	}

	errors = 0;
	warnings = 0;
	for (int i = 2 + skipargs; i < argc; i++)
		errors += process_image(argv[i], "");

	fprintf(stderr, "Processed %d file(s) with %d error(s) and %d warning(s).\n",
			argc - 2, errors, warnings);
	free_pixdata();
	return 0;
}
